plugins/catalog-backend-module-gitlab/src/providers/GitlabOrgDiscoveryEntityProvider.ts

```diff
@@ -321,6 +478,304 @@ export class GitlabOrgDiscoveryEntityProvider implements EntityProvider {
+      await this.userTransformer({
+        user: user,
+        integrationConfig: this.integration.config,
+        providerConfig: this.config,
+        groupNameTransformer: this.groupNameTransformer,
+      }),
+    );
+
+    const { added, removed } = createDeltaOperation([...userEntities]);
+
+    this.logger.debug(`Applying mutation for user ${user.username}.`);
+    await this.connection.applyMutation({
+      type: 'delta',
+      removed,
+      added,
+    });
+  }
+
+  // the goal here is to reconstruct the group either from which the user was removed or to which the user was added. Specifically, we add/remove the new user to/from the spec.member property array of the group entity. The Processor should take care of updating the relations
+  private async onMembershipChange(
+    event: SystemHookCreateOrDestroyMembershipEventsSchema,
+    createDeltaOperation: (entities: Entity[]) => {
+      added: any[];
+      removed: any[];
+    },
+  ): Promise<void> {
+    if (!this.connection) {
+      throw new Error(
+        `Gitlab discovery connection not initialized for ${this.getProviderName()}`,
+      );
+    }
+
+    const client = new GitLabClient({
+      config: this.integration.config,
+      logger: this.logger,
+    });
+
+    const groupToRebuild: GitLabGroup = await client.getGroupById(
+      event.group_id,
+    );
+
+    // If the group is outside the scope there is no point creating anything related to it.
+    if (!this.shouldProcessGroup(groupToRebuild)) {
+      this.logger.debug(`Skipped group ${groupToRebuild.full_path}.`);
+      return;
+    }
+
+    // get members of the group
+    const usersToBeAdded: string[] = [];
+    let groupMembers: PagedResponse<GitLabUser> = { items: [] };
+
+    const relations: string[] =
+      this.config.allowInherited === true && this.config.allowInherited
+        ? ['DIRECT', 'INHERITED']
+        : ['DIRECT'];
+
+    groupMembers = await client.getGroupMembers(
+      groupToRebuild.full_path,
+      relations,
+    );
+
+    if (groupMembers.items.length !== 0) {
+      groupMembers.items.forEach(element => {
+        if (
+          event.event_name === 'user_remove_from_group' &&
+          element.username === event.user_username
+        ) {
+          return;
+        }
+
+        usersToBeAdded.push(element.username);
+      });
+    }
+
+    const groupEntityToModify = await this.groupEntitiesTransformer({
+      groups: [groupToRebuild],
+      providerConfig: this.config,
+      groupNameTransformer: this.groupNameTransformer,
+    });
+
+    // we need to fetch the parent group's object because its representation might be changed by the groupTransformer
+    if (groupToRebuild.parent_id) {
+      const parentGroup = await client.getGroupById(groupToRebuild.parent_id);
+
+      groupEntityToModify[0].spec.parent = this.groupNameTransformer({
+        group: parentGroup,
+        providerConfig: this.config,
+      });
+    }
+
+    if (usersToBeAdded.length !== 0) {
```

Comment:
if we receive a `user_remove_from_group` event and `getGroupMembers()` returns empty array we don't modify spec.members. Is it correct?

Reply:
Hi, yes, this is correct!

Reply:
Hi, 
yeah, you're right. Again! :)
Is it better now? 

```typescript
    // update members of the group entity
    groupEntityToModify[0].spec.members =
      usersToBeAdded.length !== 0 ? usersToBeAdded.map(e => e.username) : [];
```

Reply:
yeah sounds good! You could also skip the check `groupEntityToModify[0].spec.members = usersToBeAdded.map(e => e.username);` but it's ok!